From 118b3bd5b2451be04564043136f526d0d1f4cae0 Mon Sep 17 00:00:00 2001
From: Hu Heyu <huheyu2@huawei.com>
Date: Thu, 5 Feb 2026 20:38:14 +0800
Subject: [PATCH] urma: introduce ipourma main module.

Offering: URMA
URMA inclusion
category: feature
bugzilla: NA

----------------------------------------------

This patch adds the main entry and defination of standard
types for the ipourma module,provides the creation and
initialization process for network devices based on ub device.

Signed-off-by: Hu Heyu <huheyu2@huawei.com>
---
 drivers/ub/urma/Makefile                     |   1 +
 drivers/ub/urma/ulp/Makefile                 |   6 +
 drivers/ub/urma/ulp/ipourma/Makefile         |  12 +
 drivers/ub/urma/ulp/ipourma/ipourma_main.c   | 370 +++++++++++++++++++
 drivers/ub/urma/ulp/ipourma/ipourma_main.h   |  20 +
 drivers/ub/urma/ulp/ipourma/ipourma_netdev.c |  44 +++
 drivers/ub/urma/ulp/ipourma/ipourma_netdev.h |  17 +
 drivers/ub/urma/ulp/ipourma/ipourma_res.c    |  19 +
 drivers/ub/urma/ulp/ipourma/ipourma_res.h    |  15 +
 drivers/ub/urma/ulp/ipourma/ipourma_types.h  | 357 ++++++++++++++++++
 drivers/ub/urma/ulp/ipourma/ipourma_ub.c     |  21 ++
 drivers/ub/urma/ulp/ipourma/ipourma_ub.h     |  14 +
 12 files changed, 896 insertions(+)
 create mode 100644 drivers/ub/urma/ulp/Makefile
 create mode 100644 drivers/ub/urma/ulp/ipourma/Makefile
 create mode 100644 drivers/ub/urma/ulp/ipourma/ipourma_main.c
 create mode 100644 drivers/ub/urma/ulp/ipourma/ipourma_main.h
 create mode 100644 drivers/ub/urma/ulp/ipourma/ipourma_netdev.c
 create mode 100644 drivers/ub/urma/ulp/ipourma/ipourma_netdev.h
 create mode 100644 drivers/ub/urma/ulp/ipourma/ipourma_res.c
 create mode 100644 drivers/ub/urma/ulp/ipourma/ipourma_res.h
 create mode 100644 drivers/ub/urma/ulp/ipourma/ipourma_types.h
 create mode 100644 drivers/ub/urma/ulp/ipourma/ipourma_ub.c
 create mode 100644 drivers/ub/urma/ulp/ipourma/ipourma_ub.h

diff --git a/drivers/ub/urma/Makefile b/drivers/ub/urma/Makefile
index b70bc0f9ef92..0916442e035a 100644
--- a/drivers/ub/urma/Makefile
+++ b/drivers/ub/urma/Makefile
@@ -7,4 +7,5 @@ obj-$(CONFIG_UB_URMA)	+= ubcore/
 obj-$(CONFIG_UB_URMA)	+= ubagg/
 
 obj-$(CONFIG_UB_URMA)	+= uburma/
+obj-$(CONFIG_UB_URMA)	+= ulp/
 
diff --git a/drivers/ub/urma/ulp/Makefile b/drivers/ub/urma/ulp/Makefile
new file mode 100644
index 000000000000..bb9b8fb607ec
--- /dev/null
+++ b/drivers/ub/urma/ulp/Makefile
@@ -0,0 +1,6 @@
+# SPDX-License-Identifier: GPL-2.0
+#
+# Makefile for the Linux kernel UB device drivers.
+#
+
+obj-$(CONFIG_UB_URMA)	+= ipourma/
diff --git a/drivers/ub/urma/ulp/ipourma/Makefile b/drivers/ub/urma/ulp/ipourma/Makefile
new file mode 100644
index 000000000000..689e688fa71a
--- /dev/null
+++ b/drivers/ub/urma/ulp/ipourma/Makefile
@@ -0,0 +1,12 @@
+# SPDX-License-Identifier: GPL-2.0
+#
+# Makefile for the Linux kernel UB device drivers.
+#
+
+ipourma-objs := \
+ipourma_main.o \
+ipourma_netdev.o \
+ipourma_res.o \
+ipourma_ub.o \
+
+obj-$(CONFIG_UB_URMA) += ipourma.o
diff --git a/drivers/ub/urma/ulp/ipourma/ipourma_main.c b/drivers/ub/urma/ulp/ipourma/ipourma_main.c
new file mode 100644
index 000000000000..076996c1add7
--- /dev/null
+++ b/drivers/ub/urma/ulp/ipourma/ipourma_main.c
@@ -0,0 +1,370 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright (c) Huawei Technologies Co., Ltd. 2025. All rights reserved.
+ *
+ * Description: ip over urma kernel module main entry
+ */
+
+#include <linux/module.h>
+#include <linux/fs.h>
+#include <linux/list.h>
+#include <linux/slab.h>
+#include <linux/stat.h>
+#include <linux/kernel.h>
+#include <linux/inet.h>
+#include <linux/netdevice.h>
+#include <linux/string.h>
+#include "ipourma_netdev.h"
+#include "ipourma_res.h"
+#include "ipourma_ub.h"
+#include "ub/urma/ubcore_api.h"
+#include "ub/urma/ubcore_uapi.h"
+#include "ipourma_main.h"
+
+STATIC int enable_test;
+module_param(enable_test, int, 0644);
+MODULE_PARM_DESC(enable_test, "Manually configure the enable bit.");
+
+static int ipourma_ubcore_add_device(struct ubcore_device *ubc_dev);
+static void ipourma_ubcore_remove_device(struct ubcore_device *ubc_dev, void *client_ctx);
+static void ipourma_unregister_netdev(struct ipourma_dev_priv *priv);
+static void ipourma_do_eid_change_handler(struct ubcore_event *event,
+					struct ubcore_event_handler *handler);
+
+struct ubcore_client g_ipourma_ubcore_client = {
+	.list_node = LIST_HEAD_INIT(g_ipourma_ubcore_client.list_node),
+	.client_name = "ipourma",
+	.add = ipourma_ubcore_add_device,
+	.remove = ipourma_ubcore_remove_device,
+};
+
+// alloc netdev for the first legal eid on ubcore_device
+// set client ctx data in ubcore
+STATIC int ipourma_ubcore_add_device(struct ubcore_device *ubc_dev)
+{
+	struct net_device *ipou_ndev = NULL;
+	struct ipourma_dev_priv *priv = NULL;
+
+	/* need to skip unsupported device */
+	if (strstr(ubc_dev->dev_name, "udma") == NULL)
+		return -EOPNOTSUPP;
+
+	ipou_ndev = ipourma_alloc_netdev(ubc_dev);
+	if (IS_ERR_OR_NULL(ipou_ndev)) {
+		pr_err("alloc netdev failed.\n");
+		return -EOPNOTSUPP;
+	}
+	priv = netdev_priv(ipou_ndev);
+	ubcore_set_client_ctx_data(ubc_dev, &g_ipourma_ubcore_client, priv);
+
+	queue_work(priv->register_wq, &priv->register_netdev);
+
+	return 0;
+}
+
+STATIC void ipourma_ubcore_remove_device(struct ubcore_device *ubc_dev, void *client_ctx)
+{
+	struct ipourma_dev_priv *priv = client_ctx;
+	struct net_device *dev;
+
+	if (IS_ERR_OR_NULL(priv))
+		return;
+
+	dev = priv->dev;
+	if (IS_ERR_OR_NULL(dev))
+		pr_err("ubcore_remove_device:priv->dev is NULL.\n");
+	else
+		ipourma_unregister_netdev(priv);
+}
+
+/*
+ * Definition of a series of functions that support debugfs.
+ */
+struct ipourma_address_iter *ipourma_address_iter_init(struct ipourma_dev_priv *priv)
+{
+	struct ipourma_address_iter *iter;
+
+	if (IS_ERR_OR_NULL(priv) || IS_ERR_OR_NULL(priv->urma_dev)) {
+		pr_err("address_iter_init: priv or priv->urma_dev is NULL\n");
+		return NULL;
+	}
+	iter = kzalloc(sizeof(*iter), GFP_KERNEL);
+	if (iter) {
+		iter->priv = priv;
+		iter->index = 0;
+	}
+	return iter;
+}
+
+int ipourma_address_iter_next(struct ipourma_address_iter *iter)
+{
+	struct ubcore_eid_table *eid_table;
+	uint32_t eid_cnt;
+
+	if (!iter || !iter->priv || !iter->priv->urma_dev) {
+		pr_err("address_iter_next: debugfs iter next NULL\n");
+		return 1;
+	}
+	eid_table = &iter->priv->urma_dev->eid_table;
+	spin_lock(&eid_table->lock);
+	eid_cnt = eid_table->eid_cnt;
+	spin_unlock(&eid_table->lock);
+	if (iter->index >= eid_cnt - 1)
+		return 1; // return 1 means stop
+
+	iter->index++;
+	return 0;
+}
+
+void ipourma_address_iter_read(struct ipourma_address_iter *iter, union ubcore_eid *eid)
+{
+	struct ubcore_eid_table *eid_table;
+	struct ubcore_eid_entry *entry;
+	uint32_t eid_cnt;
+
+	if (!iter || !iter->priv || !iter->priv->urma_dev || !eid) {
+		if (eid != NULL)
+			memset(eid, 0, UBCORE_EID_SIZE);
+		pr_err("address_iter_read: ptr NULL\n");
+		return;
+	}
+	eid_table = &iter->priv->urma_dev->eid_table;
+	spin_lock(&eid_table->lock);
+	eid_cnt = eid_table->eid_cnt;
+
+	if (iter->index < 0 || iter->index >= eid_cnt || !eid_table->eid_entries) {
+		spin_unlock(&eid_table->lock);
+		pr_err("address_iter_read: invalid index %d (address_cnt=%d)\n",
+			iter->index, eid_cnt);
+		return;
+	}
+	entry = &eid_table->eid_entries[iter->index];
+	if (entry->valid)
+		memcpy(eid->raw, entry->eid.raw, sizeof(entry->eid.raw));
+	else
+		memset(eid, 0, UBCORE_EID_SIZE);
+	spin_unlock(&eid_table->lock);
+}
+
+STATIC int ipourma_netdev_event(struct notifier_block *nb, unsigned long event, void *ptr)
+{
+	struct net_device *dev = netdev_notifier_info_to_dev(ptr);
+	struct ipourma_dev_priv *priv = NULL;
+
+	if (!strstr(dev->name, "ipourma"))
+		return NOTIFY_DONE;
+	priv = (struct ipourma_dev_priv *)netdev_priv(dev);
+
+	switch (event) {
+	case NETDEV_REGISTER:
+		/* create debugfs files here */
+		break;
+	case NETDEV_UNREGISTER:
+		/* create debugfs files here */
+		break;
+	}
+	return NOTIFY_DONE;
+}
+
+STATIC struct notifier_block ipourma_netdev_nb = {
+	.notifier_call = ipourma_netdev_event,
+};
+
+STATIC int ipourma_priv_eid_init(struct net_device *dev)
+{
+	struct ipourma_dev_priv *priv = netdev_priv(dev);
+	struct ubcore_device *urma_dev = priv->urma_dev;
+	uint32_t eid_cnt;
+	uint32_t i, j = 0;
+
+	priv->eid_info = kcalloc(IPOURMA_MAX_EID_CNT,
+		sizeof(struct ubcore_eid_info), GFP_KERNEL);
+	if (IS_ERR_OR_NULL(priv->eid_info)) {
+		priv->eid_count = 0;
+		pr_err("eid_info create failed.\n");
+		return -1;
+	}
+	priv->tx_ring_is_full = kcalloc(IPOURMA_MAX_EID_CNT, sizeof(bool), GFP_KERNEL);
+	if (IS_ERR_OR_NULL(priv->tx_ring_is_full)) {
+		kfree(priv->eid_info);
+		priv->eid_count = 0;
+		pr_err("tx_ring_is_full create failed.\n");
+		return -1;
+	}
+
+	priv->eid_change_handler.event_callback = ipourma_do_eid_change_handler;
+	ubcore_register_event_handler(priv->urma_dev, &priv->eid_change_handler);
+
+	spin_lock(&urma_dev->eid_table.lock);
+	if (IS_ERR_OR_NULL(urma_dev->eid_table.eid_entries))
+		goto eid_table_unlock;
+	eid_cnt = urma_dev->eid_table.eid_cnt;
+	for (i = 0; i < eid_cnt; i++) {
+		if (eid_is_empty(&urma_dev->eid_table.eid_entries[i].eid))
+			continue;
+		priv->eid_count++;
+	}
+
+	for (i = 0, j = 0; i < eid_cnt && j < priv->eid_count; i++) {
+		if (eid_is_empty(&urma_dev->eid_table.eid_entries[i].eid)
+			|| urma_dev->eid_table.eid_entries[i].valid == false)
+			continue;
+		pr_debug("    eid:"EID_FMT"\n",
+			EID_ARGS(urma_dev->eid_table.eid_entries[i].eid));
+
+		priv->eid_info[i].eid = urma_dev->eid_table.eid_entries[i].eid;
+		priv->eid_info[i].eid_index = i;
+		priv->eid_info_exist[j++] = priv->eid_info[i];
+	}
+	if (j != priv->eid_count)
+		pr_err("eid count error\n");
+	pr_debug("eid_count:%d\n", priv->eid_count);
+
+eid_table_unlock:
+	spin_unlock(&urma_dev->eid_table.lock);
+	return IPOURMA_OK;
+}
+
+STATIC void ipourma_cleanup_res(struct net_device *dev)
+{
+	if (IS_ERR_OR_NULL(dev))
+		return;
+	ipourma_urma_dev_uninit(dev);
+}
+
+STATIC void ipourma_proc_eid_exist(struct ipourma_dev_priv *priv)
+{
+	for (uint32_t i = 0; i < priv->eid_count; i++)
+		ipourma_create_new_eid(priv, priv->eid_info_exist[i].eid_index);
+}
+
+/* register netdev */
+void ipourma_register_netdev(struct work_struct *work)
+{
+	struct ipourma_dev_priv *priv;
+	int ret = 0;
+
+	priv = container_of(work, struct ipourma_dev_priv, register_netdev);
+	ret = register_netdev(priv->dev);
+	if (ret != 0) {
+		pr_err("register %s failed, ret = %d\n", priv->dev->name, ret);
+		goto REG_ERR;
+	}
+	if (IS_ERR_OR_NULL(priv->net_config_wq)) {
+		pr_err("%s create work queue failed, ret = %d\n", priv->dev->name, ret);
+		goto ALLOC_RESOURCE_ERR;
+	}
+	ret = ipourma_priv_eid_init(priv->dev);
+	if (ret != IPOURMA_OK) {
+		pr_err("%s init eid resources failed, ret = %d\n", priv->dev->name, ret);
+		goto ALLOC_RESOURCE_ERR;
+	}
+	ipourma_proc_eid_exist(priv);
+	queue_work(priv->net_config_wq, &(priv->set_dev_up));
+
+	// ipourma_register_sysfs(priv);
+	return;
+
+ALLOC_RESOURCE_ERR:
+	unregister_netdev(priv->dev);
+	ipourma_cleanup_res(priv->dev);
+REG_ERR:
+	return;
+}
+
+STATIC inline void ipourma_flush_register_wq(struct ipourma_dev_priv *priv)
+{
+	if (!IS_ERR_OR_NULL(priv->register_wq)) {
+		flush_workqueue(priv->register_wq);
+		destroy_workqueue(priv->register_wq);
+	}
+}
+
+STATIC void ipourma_unregister_netdev(struct ipourma_dev_priv *priv)
+{
+	ubcore_unregister_event_handler(priv->urma_dev, &priv->eid_change_handler);
+	ipourma_flush_register_wq(priv);
+	// ipourma_unregister_sysfs(priv);
+	unregister_netdev(priv->dev);
+	ipourma_reset_rings(priv);
+	ipourma_cleanup_res(priv->dev);
+	ipourma_unalloc_netdev(priv->dev);
+}
+
+STATIC void ipourma_do_eid_change_handler(struct ubcore_event *event,
+					struct ubcore_event_handler *handler)
+{
+	if (event->event_type != UBCORE_EVENT_EID_CHANGE)
+		return;
+
+	struct ubcore_device *ub_dev = event->ub_dev;
+	uint32_t eid_idx = event->element.eid_idx;
+	struct ipourma_dev_priv *priv;
+	union ubcore_eid eid;
+	uint32_t eid_idx_tmp;
+	uint32_t eid_cnt = 0;
+
+	spin_lock(&ub_dev->eid_table.lock);
+	eid_cnt = ub_dev->eid_table.eid_cnt;
+	if (eid_idx >= eid_cnt ||
+		eid_idx >= IPOURMA_MAX_EID_CNT ||
+		IS_ERR_OR_NULL(ub_dev->eid_table.eid_entries))
+		goto unlock_table_out;
+	eid_idx_tmp = ub_dev->eid_table.eid_entries[eid_idx].eid_index;
+	if (eid_idx_tmp != event->element.eid_idx ||
+		ub_dev->eid_table.eid_entries[eid_idx].valid == false)
+		goto unlock_table_out;
+	eid = ub_dev->eid_table.eid_entries[eid_idx].eid;
+	spin_unlock(&ub_dev->eid_table.lock);
+
+	priv = ubcore_get_client_ctx_data(ub_dev, &g_ipourma_ubcore_client);
+	if (!IS_ERR_OR_NULL(priv)) {
+		if (!eid_is_empty(&priv->eid_info[eid_idx].eid)) {
+			netdev_dbg(priv->dev, "get an old eid. eid index:%d, eid:"EID_FMT"\n",
+			eid_idx, EID_ARGS(priv->eid_info[eid_idx].eid));
+			return;
+		}
+		priv->eid_info[eid_idx].eid = eid;
+		priv->eid_info[eid_idx].eid_index = eid_idx;
+		netdev_dbg(priv->dev, " get a new eid. eid index:%d, eid:"EID_FMT"\n",
+			eid_idx, EID_ARGS(priv->eid_info[eid_idx].eid));
+		atomic_add(1, &(priv->need_set_ip_route));
+		ipourma_create_new_eid(priv, eid_idx);
+		priv->eid_count++;
+	}
+	return;
+unlock_table_out:
+	spin_unlock(&ub_dev->eid_table.lock);
+}
+
+STATIC int __init ipourma_init(void)
+{
+	int ret = 0;
+
+	ret = ubcore_register_client(&g_ipourma_ubcore_client);
+	if (ret != 0) {
+		pr_err("Register ubcore client failed.\n");
+		return ret;
+	}
+	pr_info("Register ubcore client success.\n");
+
+	register_netdevice_notifier(&ipourma_netdev_nb);
+
+	return 0;
+}
+
+
+STATIC void __exit ipourma_exit(void)
+{
+	ubcore_unregister_client(&g_ipourma_ubcore_client);
+	unregister_netdevice_notifier(&ipourma_netdev_nb);
+}
+#if !defined(CONFIG_IPOURMA_TEST) && !defined(CONFIG_IPOURMA_KFUZZ)
+
+module_init(ipourma_init);
+module_exit(ipourma_exit);
+
+MODULE_DESCRIPTION("ip over urma kernel module");
+MODULE_AUTHOR("huawei");
+MODULE_LICENSE("GPL");
+#endif
diff --git a/drivers/ub/urma/ulp/ipourma/ipourma_main.h b/drivers/ub/urma/ulp/ipourma/ipourma_main.h
new file mode 100644
index 000000000000..f8382f06105e
--- /dev/null
+++ b/drivers/ub/urma/ulp/ipourma/ipourma_main.h
@@ -0,0 +1,20 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/*
+ * Copyright (c) Huawei Technologies Co., Ltd. 2021-2025. All rights reserved.
+ *
+ * Description: Functions definition of ipourma_main
+ */
+
+#ifndef _IPOURMA_MAIN_H
+#define _IPOURMA_MAIN_H
+
+#include "ipourma_types.h"
+
+extern struct ubcore_client g_ipourma_ubcore_client;
+
+void ipourma_register_netdev(struct work_struct *work);
+struct ipourma_address_iter *ipourma_address_iter_init(struct ipourma_dev_priv *priv);
+int ipourma_address_iter_next(struct ipourma_address_iter *iter);
+void ipourma_address_iter_read(struct ipourma_address_iter *iter, union ubcore_eid *eid);
+
+#endif
diff --git a/drivers/ub/urma/ulp/ipourma/ipourma_netdev.c b/drivers/ub/urma/ulp/ipourma/ipourma_netdev.c
new file mode 100644
index 000000000000..3e7f1c0495f7
--- /dev/null
+++ b/drivers/ub/urma/ulp/ipourma/ipourma_netdev.c
@@ -0,0 +1,44 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright (c) Huawei Technologies Co., Ltd. 2025. All rights reserved.
+ *
+ * Description: ipourma netdev operations
+ */
+
+#include <net/addrconf.h>
+#include <net/pkt_sched.h>
+#include <net/ip_fib.h>
+#include <net/route.h>
+#include <linux/version.h>
+#include <linux/netdevice.h>
+#include <linux/ipv6.h>
+#include <linux/in6.h>
+#include <linux/if_ether.h>
+#include <linux/if_arp.h>
+#include <linux/interrupt.h>
+#include <linux/netlink.h>
+#include <linux/rtnetlink.h>
+#include <linux/ethtool.h>
+
+#include "ipourma_res.h"
+#include "ipourma_ub.h"
+#include "ipourma_main.h"
+#include "ub/urma/ubcore_uapi.h"
+#include "ipourma_netdev.h"
+
+#define IN6_ADDR_BITS   (sizeof(struct in6_addr) * 8)
+
+struct net_device *ipourma_alloc_netdev(struct ubcore_device *urma_dev)
+{
+	return NULL;
+}
+
+int ipourma_unalloc_netdev(struct net_device *dev)
+{
+	return 0;
+}
+
+void ipourma_create_new_eid(struct ipourma_dev_priv *priv, u32 eid_idx)
+{
+	/* this function will be filled in the next commit */
+}
diff --git a/drivers/ub/urma/ulp/ipourma/ipourma_netdev.h b/drivers/ub/urma/ulp/ipourma/ipourma_netdev.h
new file mode 100644
index 000000000000..a329188585ac
--- /dev/null
+++ b/drivers/ub/urma/ulp/ipourma/ipourma_netdev.h
@@ -0,0 +1,17 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/*
+ * Copyright (c) Huawei Technologies Co., Ltd. 2021-2025. All rights reserved.
+ *
+ * Description: Functions definition of ipourma_netdev
+ */
+
+#ifndef _IPOURMA_NETDEV_H
+#define _IPOURMA_NETDEV_H
+
+#include "ipourma_types.h"
+
+int ipourma_unalloc_netdev(struct net_device *dev);
+struct net_device *ipourma_alloc_netdev(struct ubcore_device *urma_dev);
+void ipourma_create_new_eid(struct ipourma_dev_priv *priv, u32 eid_idx);
+
+#endif
diff --git a/drivers/ub/urma/ulp/ipourma/ipourma_res.c b/drivers/ub/urma/ulp/ipourma/ipourma_res.c
new file mode 100644
index 000000000000..5d3096031d49
--- /dev/null
+++ b/drivers/ub/urma/ulp/ipourma/ipourma_res.c
@@ -0,0 +1,19 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright (c) Huawei Technologies Co., Ltd. 2025. All rights reserved.
+ *
+ * Description: ipourma urma resource management
+ */
+#include <linux/timer.h>
+#include "ipourma_ub.h"
+#include "ub/urma/ubcore_uapi.h"
+#include "ipourma_res.h"
+
+u32 ipourma_tx_ring_size __read_mostly = IPOURMA_TX_RING_SIZE;
+u32 ipourma_rx_ring_size __read_mostly = IPOURMA_RX_RING_SIZE;
+
+int ipourma_reset_rings(struct ipourma_dev_priv *priv)
+{
+	return IPOURMA_OK;
+}
+
diff --git a/drivers/ub/urma/ulp/ipourma/ipourma_res.h b/drivers/ub/urma/ulp/ipourma/ipourma_res.h
new file mode 100644
index 000000000000..e1787d2babe0
--- /dev/null
+++ b/drivers/ub/urma/ulp/ipourma/ipourma_res.h
@@ -0,0 +1,15 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/*
+ * Copyright (c) Huawei Technologies Co., Ltd. 2021-2025. All rights reserved.
+ *
+ * Description: Functions definition of ipourma_res
+ */
+
+#ifndef _IPOURMA_RES_H
+#define _IPOURMA_RES_H
+
+#include "ipourma_types.h"
+
+int ipourma_reset_rings(struct ipourma_dev_priv *priv);
+
+#endif
diff --git a/drivers/ub/urma/ulp/ipourma/ipourma_types.h b/drivers/ub/urma/ulp/ipourma/ipourma_types.h
new file mode 100644
index 000000000000..af26a7bf86cc
--- /dev/null
+++ b/drivers/ub/urma/ulp/ipourma/ipourma_types.h
@@ -0,0 +1,357 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/*
+ * Copyright (c) Huawei Technologies Co., Ltd. 2021-2025. All rights reserved.
+ *
+ * Description: Types definition for ipourma
+ */
+
+#ifndef _IPOURMA_TYPES_H
+#define _IPOURMA_TYPES_H
+
+#include <linux/netdevice.h>
+#include <linux/skbuff.h>
+#include <linux/jhash.h>
+#include <linux/list.h>
+#include <linux/slab.h>
+#include <linux/random.h>
+#include <linux/timer.h>
+#include "ub/urma/ubcore_types.h"
+
+#if defined(CONFIG_IPOURMA_TEST) || defined(CONFIG_IPOURMA_KFUZZ)
+#define STATIC
+#else
+#define STATIC static
+#endif
+
+static inline int eid_is_empty(union ubcore_eid *eid)
+{
+	return (eid->in6.interface_id == 0 && eid->in6.subnet_prefix == 0);
+}
+
+static inline bool ipourma_are_eids_equal(union ubcore_eid *eid1,
+	union ubcore_eid *eid2)
+{
+	if (memcmp(eid1, eid2, sizeof(union ubcore_eid)) == 0)
+		return true;
+	return false;
+}
+
+/*
+ * It's controlled by ipourma itself, no need to comply with any public protocol
+ * For now, we just use "ether type" as ipourma link header
+ */
+struct ipourma_header {
+	__be16 proto;
+};
+
+enum {
+	IPOURMA_DEV_ADMIN_UP        = 1,
+	IPOURMA_DEV_OP_UP           = 2,
+	IPOURMA_JFS_DEPTH           = 128,
+	IPOURMA_JFR_DEPTH           = 256,
+	IPOURMA_TX_JFC_DEPTH        = 1024,
+	IPOURMA_RX_JFC_DEPTH        = 2048,
+	IPOURMA_SEGMENT_ALIGN_SIZE  = 4096,
+	IPOURMA_REGISTER_SEG_SIZE   = 65536,
+	IPOURMA_MAX_URMA_SEND_SGES  = 13,
+	IPOURMA_MAX_URMA_RECV_SGES  = 1,
+	IPOURMA_WELL_KNOWN_JETTY_ID = 32,
+	IPOURMA_MAX_RX_SGES         = 1,
+	IPOURMA_TJETTY_HMAP_SIZE    = 1024,
+	/* additional 1 for linear data, normally 18 */
+	IPOURMA_MAX_TX_SGES         = MAX_SKB_FRAGS + 1,
+	IPOURMA_NAPI_RX_WEIGHT      = 4,
+	IPOURMA_NAPI_TX_WEIGHT      = 16,
+	IPOURMA_TX_RING_SIZE        = 128,
+	IPOURMA_RX_RING_SIZE        = 256,
+	IPOURMA_URMA_MAX_MTU        = 4096,
+	IPOURMA_MAX_MTU             = (IPOURMA_URMA_MAX_MTU -
+					sizeof(struct ipourma_header)),
+	IPOURMA_MIN_MTU             = 1280,
+	IPOURMA_DEFAULT_MTU         = IPOURMA_MAX_MTU,
+	IPOURMA_MIN_SKB_LEN         = 68,
+	IPOURMA_ALEN                = 6,
+	IPOURMA_DEFAULT_TJETTY_CAP  = 256,
+	IPOURMA_MAX_EID_CNT         = 32,
+	IPOURMA_TJETTY_CB_S         = 10,
+	IPOURMA_TJETTY_TIMEOUT_S    = 60,
+	IPOURMA_TJETTY_TIMEOUT_MAX  = 65535,
+	IPOURMA_MAX_DEV_NAME        = 50,
+};
+
+enum {
+	IPOURMA_OK = 0,
+	IPOURMA_INVALID_IPV6_ADDR,
+	IPOURMA_UNSUPPORTED_ETH_PROTO,
+	IPOURMA_ALLOC_NETDEV_FAILED,
+	IPOURMA_ALLOC_DEV_PRIV_FAILED,
+	IPOURMA_ALLOC_TX_RING_TABLE_FAILED,
+	IPOURMA_ALLOC_RX_RING_TABLE_FAILED,
+	IPOURMA_ALLOC_TX_RING_FAILED,
+	IPOURMA_ALLOC_RX_RING_FAILED,
+	IPOURMA_TX_RING_FULL,
+	IPOURMA_REGISTER_SEG_FAILED,
+	IPOURMA_INIT_URMA_RES_FAILED,
+	IPOURMA_POST_SEND_FAILED,
+	IPOURMA_POST_RECV_FAILED,
+	IPOURMA_ALLOC_RX_SKB_FAILED,
+	IPOURMA_URMA_POST_RECV_FAILED,
+	IPOURMA_REARM_JFC_FAILED,
+	IPOURMA_INCORRECT_CR_STATUS,
+	IPOURMA_INCORRECT_WQE_JETTY_IDX,
+	IPOURMA_INCORRECT_WQE_IDX,
+	IPOURMA_POLL_JFC_FAILED,
+	IPOURMA_CREATE_JFR_TABLE_FAILED,
+	IPOURMA_CREATE_JETTY_TABLE_FAILED,
+	IPOURMA_CREATE_JFC_FAILED,
+	IPOURMA_CREATE_JFR_FAILED,
+	IPOURMA_CREATE_JETTY_FAILED,
+	IPOURMA_SRC_IP_ADDR_EID_MISMATCH,
+	IPOURMA_NOT_SUPPORT_GSO,
+	IPOURMA_GIANT_PACKET,
+	IPOURMA_TOO_MANY_FRAGS,
+	IPOURMA_IMPORT_JETTY_FAILED,
+	IPOURMA_TX_CQE_ERR,
+	IPOURMA_INSUFFICIENT_MEMORY,
+	IPOURMA_ADDRESS_NOT_ALIGNED,
+	IPOURMA_UNSUPPORTED_LINEAR_DATA_LEN,
+	IPOURMA_REPLENISH_RX_SEG_FAILED,
+	IPOURMA_NLMSG_ERR,
+	IPOURMA_NLDATA_ERR,
+	IPOURMA_NL_SEND_ERR,
+	IPOURMA_EXCEED_SKB_LENGTH,
+	IPOURMA_INIT_TJETTY_HMAP_FAILED,
+	IPOURMA_ALLOC_TX_RING_LOCKS_FAILED,
+	IPOURMA_TJETTY_NODE_ALLOC_FAILED,
+	IPOURMA_ALLOC_CR_FAILED,
+	IPOURMA_FLUSH_JETTY_FAILED,
+	IPOURMA_MODIFY_JETTY_FAILED,
+	IPOURMA_MODIFY_JFR_FAILED,
+	IPOURMA_INIT_RINGS_TABLE_FAILED,
+	IPOURMA_MAX_ERRNO
+};
+
+enum {
+	IFLA_IPOURMA_UNSPEC,
+	IFLA_IPOURMA_OP_MODE,
+	IFLA_IPOURMA_TRANSPORT_MODE,
+	IFLA_IPOURMA_MAX_SEND_SGE,
+	IFLA_IPOURMA_XMTU,
+	__IFLA_IPOURMA_MAX,
+};
+
+#define IFLA_IPOURMA_MAX (__IFLA_IPOURMA_MAX - 1)
+extern u32 ipourma_tx_ring_size;
+extern u32 ipourma_rx_ring_size;
+
+/* IPOURMA_MAX_CR_STATUS should be as same as the length of the enum ubcore_cr_status. Since
+ * ubcore_cr_status is contiguous and starts from 0, we can use the value of the last item
+ * plus one to represent the length of ubcore_cr_status.
+ */
+
+#define IPOURMA_MAX_CR_STATUS (UBCORE_CR_REM_DATA_POISON + 1)
+
+struct ipourma_tx_stats {
+	/* start xmit */
+	uint64_t num_recv_pkts_from_kernel;
+	uint64_t post_send_enque;
+	uint64_t post_send_bypass;
+	uint64_t gso_not_support;
+	uint64_t packet_size_error;
+	uint64_t frag_error;
+	uint64_t not_ipv6_proto;
+	uint64_t not_ipv6_addr;
+	uint64_t ip_eid_not_equal;
+	uint64_t tx_ring_full;
+	/* post send */
+	uint64_t post_send_start;
+	uint64_t num_import_jetty_real;
+	uint64_t num_import_jetty_bypass;
+	uint64_t num_tjetty_hash_hit;
+	uint64_t linear_len_oversize;
+	uint64_t import_jetty_failed;
+	uint64_t send_wr_failed;
+	/* pass to ub */
+	uint64_t pass_to_ub;
+	/* tx cqe notify */
+	uint64_t cqe_notify;
+	/* tx poll */
+	uint64_t num_napi_tx;
+	uint64_t cqe_recved;
+	uint64_t cqe_success;
+	uint64_t cqe_err;
+	uint64_t cqe_stats[IPOURMA_MAX_CR_STATUS];
+	uint64_t poll_jfc_success;
+	uint64_t poll_jfc_failed;
+	uint64_t rearm_success;
+	uint64_t rearm_failed;
+	uint64_t flush_jetty_success;
+};
+
+struct ipourma_rx_stats {
+	/* rx cqe notify */
+	uint64_t cqe_notify;
+	/* rx poll */
+	uint64_t num_napi_rx;
+	uint64_t rx_enque;
+	uint64_t rx_deque;
+	uint64_t cqe_recved;
+	uint64_t cqe_success;
+	uint64_t cqe_err;
+	uint64_t cqe_stats[IPOURMA_MAX_CR_STATUS];
+	uint64_t poll_jfc_success;
+	uint64_t poll_jfc_failed;
+	uint64_t rearm_success;
+	uint64_t rearm_failed;
+	uint64_t cr_len_err;
+	uint64_t replenish_enque;
+	/* pass to kernel */
+	uint64_t pass_to_kernel;
+	/* replenish */
+	uint64_t replenish_deque;
+	uint64_t num_post_wr;
+	uint64_t alloc_skb_failed;
+	uint64_t register_seg_failed;
+	uint64_t post_wr_failed;
+};
+
+struct ipourma_runtime_stats {
+	struct ipourma_tx_stats tx_stats;
+	struct ipourma_rx_stats rx_stats;
+	spinlock_t lock;
+};
+
+struct ipourma_rx_buf {
+	struct ipourma_dev_priv *priv;
+	struct sk_buff *skb_pass_up;
+	u8 *buf_aligned;
+	struct ubcore_target_seg *seg[IPOURMA_MAX_RX_SGES];
+	u32 idx;
+	/* eid_index = jetty id - IPOURMA_WELL_KNOWN_JETTY_ID */
+	u32 eid_index;
+	struct work_struct work;
+	struct ubcore_jfr_wr rx_wr;
+	struct ubcore_sge rx_sge[IPOURMA_MAX_RX_SGES];
+};
+
+struct ipourma_tx_buf {
+	struct ipourma_dev_priv *priv;
+	/* for skb linear data */
+	u8 *buf_aligned;
+	struct ubcore_target_seg *seg[IPOURMA_MAX_TX_SGES];
+	u32 idx;
+	u32 eid_index;
+	struct work_struct work;
+	/* dynamic fields */
+	struct sk_buff *skb;
+	union ubcore_eid dst_eid;
+	u8 tx_buf_in_use;
+	struct ubcore_jfs_wr tx_wr;
+	struct ubcore_sge tx_sge[IPOURMA_MAX_TX_SGES];
+	struct ubcore_tjetty *tjetty;
+};
+
+struct ipourma_ethtool_st {
+	u16     coalesce_usecs;
+	u16     max_coalesced_frames;
+};
+
+struct ipourma_tjetty_hash_node {
+	struct hlist_node hlist;
+	struct list_head lru_list;
+	union ubcore_eid key[2];
+	struct ubcore_tjetty *tjetty;
+	u64 last_jiffies;
+};
+
+struct ipourma_tjetty_hmap {
+	struct hlist_head *buckets;
+	u32 hash_seed;
+};
+
+struct ipourma_tjetty_lru {
+	struct ipourma_tjetty_hmap tjetty_hmap;
+	struct list_head list;
+	spinlock_t lock;
+	uint16_t count;
+	uint16_t tjetty_capacity;
+	struct delayed_work tjetty_aging_work;
+	uint16_t tjetty_aging_interval_s;
+	uint16_t tjetty_aging_timeout_s;
+};
+
+struct ipourma_dev_priv {
+	struct net_device *dev;
+	struct net_device *parent;
+	/* ipourma device */
+	struct ubcore_device *urma_dev;
+	struct ipourma_tx_buf **tx_ring;
+	struct ipourma_rx_buf **rx_ring;
+	u32 tx_ring_size;
+	u32 rx_ring_size;
+	u32 tx_buf_size;
+	u32 *tx_head;
+	u32 *tx_tail;
+	u32 *tx_count;
+	struct napi_struct napi_send;
+	struct napi_struct napi_recv;
+	struct workqueue_struct *tx_wq;
+	struct workqueue_struct *rx_wq;
+	unsigned long flags;
+	struct ipourma_ethtool_st ethtool;
+	u8 route_tbl_idx;
+	spinlock_t lock;
+	spinlock_t *tx_ring_locks;
+	struct dentry *address_dentry;
+	atomic_t need_set_ip_route;
+	struct workqueue_struct *net_config_wq;
+	struct work_struct set_dev_up;
+	struct work_struct set_ip;
+	struct work_struct set_route;
+	struct work_struct unset_route;
+	struct work_struct set_route_entry;
+	struct work_struct rx_cr_event;
+	/* register netdev */
+	struct workqueue_struct *register_wq;
+	struct work_struct register_netdev;
+	/* urma device */
+	struct ubcore_eid_info *eid_info;
+	struct ubcore_eid_info eid_info_exist[IPOURMA_MAX_EID_CNT];
+	uint32_t eid_count;
+	struct ubcore_jetty **jetty;
+	struct ubcore_jfc *tx_jfc;
+	struct ubcore_jfc *rx_jfc;
+	struct ubcore_jfr **jfr;
+	struct ubcore_cr tx_cr[IPOURMA_NAPI_TX_WEIGHT];
+	struct ubcore_cr rx_cr[IPOURMA_NAPI_RX_WEIGHT];
+	enum ubcore_opcode urma_op_mode;
+	enum ubcore_transport_mode urma_transport_mode;
+	u32 max_send_sge;
+	u32 urma_mtu;
+	struct ubcore_event_handler eid_change_handler;
+	atomic_t rx_jfr_ref;
+	bool *tx_ring_is_full;
+	atomic_t tx_ring_blocked;
+	/* tjetty lru */
+	struct ipourma_tjetty_lru tjetty_lru;
+	/* runtime stats statistics */
+	struct ipourma_runtime_stats runtime_stats;
+	/* memory pool */
+	struct ubcore_target_seg **ipourma_ub_tx_seg[IPOURMA_MAX_EID_CNT];
+	struct ubcore_target_seg **ipourma_ub_rx_seg[IPOURMA_MAX_EID_CNT];
+	u8 **tx_buf_aligned[IPOURMA_MAX_EID_CNT];
+	u8 **rx_buf_aligned[IPOURMA_MAX_EID_CNT];
+	size_t tx_buf_num;
+	size_t rx_buf_num;
+	u32 skb_buf_size;
+};
+
+/* iterator for debugfs-eid */
+struct ipourma_address_iter {
+	struct ipourma_dev_priv *priv;
+	uint32_t index;
+};
+
+#define IPOURMA_HARD_LEN ((unsigned short)sizeof(struct ipourma_header))
+
+#endif
diff --git a/drivers/ub/urma/ulp/ipourma/ipourma_ub.c b/drivers/ub/urma/ulp/ipourma/ipourma_ub.c
new file mode 100644
index 000000000000..6b53ec8ad2c5
--- /dev/null
+++ b/drivers/ub/urma/ulp/ipourma/ipourma_ub.c
@@ -0,0 +1,21 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright (c) Huawei Technologies Co., Ltd. 2025. All rights reserved.
+ *
+ * Description: ipourma underlay datapath implementaion
+ */
+
+#include <net/dst.h>
+#include <linux/version.h>
+#include <linux/types.h>
+#include <linux/smp.h>
+#include <linux/jiffies.h>
+#include "ipourma_res.h"
+#include "ipourma_main.h"
+#include "ub/urma/ubcore_uapi.h"
+#include "ipourma_ub.h"
+
+void ipourma_urma_dev_uninit(struct net_device *dev)
+{
+	/* this function will be filled in the next commit */
+}
diff --git a/drivers/ub/urma/ulp/ipourma/ipourma_ub.h b/drivers/ub/urma/ulp/ipourma/ipourma_ub.h
new file mode 100644
index 000000000000..ffd4dbc549dc
--- /dev/null
+++ b/drivers/ub/urma/ulp/ipourma/ipourma_ub.h
@@ -0,0 +1,14 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/*
+ * Copyright (c) Huawei Technologies Co., Ltd. 2021-2025. All rights reserved.
+ *
+ * Description: Functions definition of ipourma_ub
+ */
+
+#ifndef _IPOURMA_UB_H
+#define _IPOURMA_UB_H
+
+#include "ipourma_types.h"
+
+void ipourma_urma_dev_uninit(struct net_device *dev);
+#endif
-- 
2.43.0

